toolchain/parser/parser_impl.cpp

```diff
@@ -533,63 +539,71 @@ auto ParseTree::Parser::ParseFunctionDeclaration() -> Node {
   TokenizedBuffer::Token function_intro_token = Consume(TokenKind::Fn());
   auto start = GetSubtreeStartPosition();
 
-  auto add_error_function_node = [&] {
+  auto add_error_function_node = [&](bool skip_past_likely_end) {
+    if (skip_past_likely_end) {
+      if (auto semi_token = SkipPastLikelyEnd(function_intro_token)) {
+        AddLeafNode(ParseNodeKind::DeclarationEnd(), *semi_token);
+      }
+    }
     return AddNode(ParseNodeKind::FunctionDeclaration(), function_intro_token,
                    start, /*has_error=*/true);
   };
-  CARBON_RETURN_IF_STACK_LIMITED(add_error_function_node());
+  CARBON_RETURN_IF_STACK_LIMITED(add_error_function_node(false));
 
-  auto handle_semi_in_error_recovery = [&](TokenizedBuffer::Token semi) {
-    return AddLeafNode(ParseNodeKind::DeclarationEnd(), semi);
-  };
-
-  auto name_n = ConsumeAndAddLeafNodeIf(TokenKind::Identifier(),
-                                        ParseNodeKind::DeclaredName());
-  if (!name_n) {
+  if (!ConsumeAndAddLeafNodeIf(TokenKind::Identifier(),
+                               ParseNodeKind::DeclaredName())) {
     CARBON_DIAGNOSTIC(ExpectedFunctionName, Error,
                       "Expected function name after `fn` keyword.");
     emitter_.Emit(*position_, ExpectedFunctionName);
     // TODO: We could change the lexer to allow us to synthesize certain
     // kinds of tokens and try to "recover" here, but unclear that this is
     // really useful.
-    SkipPastLikelyEnd(function_intro_token, handle_semi_in_error_recovery);
-    return add_error_function_node();
+    return add_error_function_node(true);
   }
 
   TokenizedBuffer::Token open_paren = *position_;
   if (tokens_.GetKind(open_paren) != TokenKind::OpenParen()) {
     CARBON_DIAGNOSTIC(ExpectedFunctionParams, Error,
                       "Expected `(` after function name.");
     emitter_.Emit(open_paren, ExpectedFunctionParams);
-    SkipPastLikelyEnd(function_intro_token, handle_semi_in_error_recovery);
-    return add_error_function_node();
+    AddNode(ParseNodeKind::FunctionSignature(), *position_, start,
+            /*has_error=*/true);
+    return add_error_function_node(true);
   }
   TokenizedBuffer::Token close_paren =
       tokens_.GetMatchedClosingToken(open_paren);
 
   if (!ParseFunctionSignature()) {
     // Don't try to parse more of the function declaration, but consume a
     // declaration ending semicolon if found (without going to a new line).
-    SkipPastLikelyEnd(function_intro_token, handle_semi_in_error_recovery);
-    return add_error_function_node();
+    AddNode(ParseNodeKind::FunctionSignature(), *position_, start,
+            /*has_error=*/true);
+    return add_error_function_node(true);
   }
 
-  // See if we should parse a definition which is represented as a code block.
-  if (NextTokenIs(TokenKind::OpenCurlyBrace())) {
-    if (!ParseCodeBlock()) {
-      return add_error_function_node();
+  AddNode(ParseNodeKind::FunctionSignature(), *position_, start);
```

Comment:
I guess this is the "virtual" parse tree node -- `*position_` (the `;` or `{`) will have two correspoding `ParseTreeNode`s, one for the `FunctionSignature` and another for the `DeclarationEnd` or `CodeBlock`. If we're going to have this refer to a token that another parse tree node also refers to, I wonder if it would be more useful for the token corresponding to the `FunctionSignature` to be the `fn` token (`function_intro_token`) instead?

Reply:
As above